archiver/service/archiver.go

```diff
@@ -0,0 +1,248 @@
+	"github.com/attestantio/go-eth2-client/api"
+	v1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/base-org/blob-archiver/archiver/flags"
+	"github.com/base-org/blob-archiver/archiver/metrics"
+	"github.com/base-org/blob-archiver/common/storage"
+	"github.com/ethereum-optimism/optimism/op-service/httputil"
+	opmetrics "github.com/ethereum-optimism/optimism/op-service/metrics"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const backfillErrorRetryInterval = 5 * time.Second
+
+var ErrAlreadyStopped = errors.New("already stopped")
+
+type BeaconClient interface {
+	client.BlobSidecarsProvider
+	client.BeaconBlockHeadersProvider
+}
+
+func NewService(l log.Logger, cfg flags.ArchiverConfig, api *API, dataStoreClient storage.DataStore, client BeaconClient, m metrics.Metricer) (*ArchiverService, error) {
+	return &ArchiverService{
+		log:             l,
+		cfg:             cfg,
+		dataStoreClient: dataStoreClient,
+		metrics:         m,
+		stopCh:          make(chan struct{}),
+		beaconClient:    client,
+		api:             api,
+	}, nil
+}
+
+type ArchiverService struct {
+	stopped         atomic.Bool
+	stopCh          chan struct{}
+	log             log.Logger
+	dataStoreClient storage.DataStore
+	beaconClient    BeaconClient
+	metricsServer   *httputil.HTTPServer
+	cfg             flags.ArchiverConfig
+	metrics         metrics.Metricer
+	api             *API
+}
+
+// Start starts the archiver service. It begins polling the beacon node for the latest blocks and persisting blobs for
+// them. Concurrently it'll also begin a backfill process (see backfillBlobs) to store all blobs from the current head
+// to the previously stored blocks. This ensures that during restarts or outages of an archiver, any gaps will be
+// filled in.
+func (a *ArchiverService) Start(ctx context.Context) error {
+	if a.cfg.MetricsConfig.Enabled {
+		a.log.Info("starting metrics server", "addr", a.cfg.MetricsConfig.ListenAddr, "port", a.cfg.MetricsConfig.ListenPort)
+		srv, err := opmetrics.StartServer(a.metrics.Registry(), a.cfg.MetricsConfig.ListenAddr, a.cfg.MetricsConfig.ListenPort)
+		if err != nil {
+			return err
+		}
+
+		a.log.Info("started metrics server", "addr", srv.Addr())
+		a.metricsServer = srv
+	}
+
+	srv, err := httputil.StartHTTPServer(a.cfg.ListenAddr, a.api.router)
+	if err != nil {
+		return fmt.Errorf("failed to start Archiver API server: %w", err)
+	}
+
+	a.log.Info("Archiver API server started", "address", srv.Addr().String())
+
+	currentBlob, _, err := a.persistBlobsForBlockToS3(ctx, "head")
+	if err != nil {
+		a.log.Error("failed to seed archiver with initial block", "err", err)
+		return err
+	}
+
+	go a.backfillBlobs(ctx, currentBlob)
+
+	return a.trackLatestBlocks(ctx)
+}
+
+// persistBlobsForBlockToS3 fetches the blobs for a given block and persists them to S3. It returns the block header
+// and a boolean indicating whether the blobs already existed in S3 and any errors that occur.
```

Comment:
what is only some of the blobs from the block were in S3?

Reply:
Ah looks like it's atomically writing all of them at once.

Reply:
That's right, all the blobs are written as a single file atomically (as json). Even the largest files are low single digit megabytes, so it didn't feel worth it to write them separately (for now).